go-swagger: fix panic under Golang 1.13

kolyshkin opened the pull request:
This is an attempt to fix go-swagger panic under Golang 1.13.

Details:
 * https://github.com/go-openapi/jsonpointer/pull/4
 * https://github.com/go-swagger/go-swagger/pull/2059

Failure:
 * https://github.com/moby/moby/pull/40028#issuecomment-538140971

kolyshkin committed:
go-swagger: fix panic

This is an attempt to fix go-swagger panic under Golang 1.13.

Details:
 * https://github.com/go-openapi/jsonpointer/pull/4
 * https://github.com/go-swagger/go-swagger/pull/2059

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah reviewed (commented):
You'll have to add a commit that changes the swagger.yaml, otherwise go-swagger won't run

thaJeztah commented:
Failing, looks like you're referencing a wrong commit;

```
error: pathspec '5793aa66d4b4112c2602c716516e24710e4adbb5 # golang-1.13-fix in kolyshkin/go-swagger' did not match any file(s) known to git.

kolyshkin commented:
> error: pathspec '5793aa66d4b4112c2602c716516e24710e4adbb5 # golang-1.13-fix in kolyshkin/go-swagger' did not match any file(s) known to git.

It's just I added a comment to an inappropriate place. Removed.

thaJeztah commented:
ah, lol you just pushed again

kolyshkin commented:
OK looks like it works:

> [2019-10-04T23:46:23.571Z] The swagger spec at "api/swagger.yaml" is valid against swagger specification 2.0

Now I'm not sure what would be the best way to have it. Temporary, we can merge this one as-is (removing the second commit of course). Longer-term, IDK.

thaJeztah commented:
> Now I'm not sure what would be the best way to have it. Temporary, we can merge this one as-is (removing the second commit of course). Longer-term, IDK.

Some thoughts;

- we don't (currently) need the new features of newer go-swagger versions, so it should be ok to keep using this old version for now
- this tool is only used locally, so if future Go updates break it again, perhaps we could consider pinning it to a fixed version of go (and/or pre-building the binaries)
    - decide if we want to fix the parts they removed in go-swagger again (see https://github.com/go-swagger/go-swagger/pull/691 for the last time it was fixed), or:
    - if we want to continue the work on https://github.com/moby/moby/pull/36714, and replace it with https://github.com/dnephin/swagger-gen-types (I just noticed that that repository doesn't have a license, so we should make sure it has one if we want to switch)

thaJeztah commented on Dockerfile:
```diff
@@ -79,10 +79,10 @@ RUN set -x \
 FROM base AS swagger
 # Install go-swagger for validating swagger.yaml
-ENV GO_SWAGGER_COMMIT c28258affb0b6251755d92489ef685af8d4ff3eb
+ENV GO_SWAGGER_COMMIT 5793aa66d4b4112c2602c716516e24710e4adbb5
```
If we decide to merge this; perhaps we should tag the commit (also wondering if the fork should live in the `moby` org, so that it's not tied to a personal account)

kolyshkin commented:
OK here's my plan
1. merge this one as a temp fix to unblock CI
2. Try making upstream go-swagger work for us
3. If the above fails (I don't want to spend too much time on it), move the fork under moby and use it.

thaJeztah commented on Dockerfile:
```diff
@@ -79,10 +79,10 @@ RUN set -x \
 RUN set -x \
 	&& export GOPATH="$(mktemp -d)" \
-	&& git clone https://github.com/go-swagger/go-swagger.git "$GOPATH/src/github.com/go-swagger/go-swagger" \
+	&& git clone https://github.com/kolyshkin/go-swagger.git "$GOPATH/src/github.com/go-swagger/go-swagger" \
```
Can you add a TODO comment, and open a tracking issue?

kolyshkin replied:
done

kolyshkin commented:
Rebased, added TODO

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
Weird; CI Hung after tests completed; https://ci.docker.com/public/job/moby/job/PR-40038/6/execution/node/195/log/?consoleFull

```
00:55:32.976  DONE 270 tests, 6 skipped in 2585.326s

...


02:00:04.686  Sending interrupt signal to process
02:00:04.715  Sending interrupt signal to process
02:00:13.230  /home/ubuntu/workspace/moby_PR-40038@tmp/durable-89316c98/script.sh: line 5:  9079 Terminated              docker run $rm -t --privileged -v "$WORKSPACE/bundles/${TEST_INTEGRATION_DEST}:/go/src/github.com/docker/docker/bundles" -v "$WORKSPACE/bundles/dynbinary-daemon:/go/src/github.com/docker/docker/bundles/dynbinary-daemon" -v "$WORKSPACE/.git:/go/src/github.com/docker/docker/.git" --name "$CONTAINER_NAME" -e KEEPBUNDLE=1 -e TESTDEBUG -e TESTFLAGS -e TEST_SKIP_INTEGRATION -e TEST_SKIP_INTEGRATION_CLI -e DOCKER_GITCOMMIT=${GIT_COMMIT} -e DOCKER_GRAPHDRIVER -e TIMEOUT -e VALIDATE_REPO=${GIT_URL} -e VALIDATE_BRANCH=${CHANGE_TARGET} docker:${GIT_COMMIT} hack/make.sh "$1" test-integration
02:00:13.230  Remaining pids to kill: []
02:00:13.235  script returned exit code 143
```
